File: BlindAssist/CameraPreviewView.m
//
//  CameraPreviewView.m
//  BlindAssist
//
//  Created by Giovanni Terlingen on 27-03-18.
//  Copyright © 2018 Giovanni Terlingen. All rights reserved.
//

#import "CameraPreviewView.h"
#import "Utils.h"

@implementation CameraPreviewView

- (void)awakeFromNib {
    [super awakeFromNib];
    
    // Observe device rotation
     [[NSNotificationCenter defaultCenter]
      addObserver:self selector:@selector(orientationChanged:)
      name:UIDeviceOrientationDidChangeNotification
      object:[UIDevice currentDevice]];
}

// Insert layer at index 0
- (void)addCaptureVideoPreviewLayer:(AVCaptureVideoPreviewLayer*) previewLayer {
    [[self previewLayer] removeFromSuperlayer];
    [[self layer] insertSublayer:previewLayer atIndex:0];
    [self setPreviewLayer:(previewLayer)];
    [self previewLayer].videoGravity = AVLayerVideoGravityResizeAspectFill;
}

- (void)layoutSubviews {
    [super layoutSubviews];
    [self previewLayer].frame = [super bounds];
}

// Change video orientation to always display video in correct orientation
- (void)orientationChanged:(NSNotification*) notification {
    [self previewLayer].connection.videoOrientation = [Utils getVideoOrientation];
}

@end

File: BlindAssist/Views/CameraPreviewView.swift
//
//  CameraPreviewView.swift
//  BlindAssist
//
//  Created by Giovanni Terlingen on 22/10/2018.
//  Copyright © 2018 Giovanni Terlingen. All rights reserved.
//

// Uses code from https://github.com/vfa-tranhv/MobileAILab-HairColor-iOS

import UIKit
import AVFoundation

final class CameraPreviewView: UIView {
    
    private weak var previewLayer: AVCaptureVideoPreviewLayer?
    
    override func awakeFromNib() {
        super.awakeFromNib()
        
        // Observe device rotation
        let selector = #selector(deviceOrientationDidChange(_:))
        UIDevice.subscribeToDeviceOrientationNotifications(self, selector: selector)
    }
    
    // Insert layer at index 0
    func addCaptureVideoPreviewLayer(_ previewLayer: AVCaptureVideoPreviewLayer) {
        self.previewLayer?.removeFromSuperlayer()
        self.layer.insertSublayer(previewLayer, at: 0)
        self.previewLayer = previewLayer
        self.previewLayer?.videoGravity = .resizeAspectFill
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        previewLayer?.frame = bounds
    }
    
    // Change video orientation to always display video in correct orientation
    @objc private func deviceOrientationDidChange(_ notification: Notification) {
        previewLayer?.connection?.videoOrientation = UIDevice.current.videoOrientation
    }
}
